### PR TITLE
TRUNK-5563: Add onset date check for conditions to ConditionValidator

### PR DESCRIPTION
**Pull request by haripriya999**
# TRUNK-5563: Add onset date check for conditions to ConditionValidator

## Description of what I changed
added check for onset dates in ConditionValidator

## Issue I worked on

see https://issues.openmrs.org/browse/TRUNK-5563

## Checklist: I completed these to help reviewers :)

- [x] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: api/src/main/java/org/openmrs/validator/ConditionValidator.java [hunk 1]
```diff
@@ -44,7 +46,7 @@ public void validate(Object object, Errors errors) {
 			throw  new IllegalArgumentException("The object parameter should be of type " + Condition.class);
 		}
 		Condition condition = (Condition) object;
-		if(condition.getCondition() == null){
+		if(condition.getCondition() == null || (condition.getClinicalStatus().equals(ConditionClinicalStatus.INACTIVE) && condition.getOnsetDate().compareTo(condition.getEndDate()) > 0)){
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/validator/ConditionValidator.java]: Don't do a OR statement here, do a separate validation under those that are there already.

**Comment by haripriya999** [comment]
@mks-d  Travis is failing due to this error
> ConditionValidatorTest.shouldFailIfGivenConditionWithNullConditionProperties:70 » NullPointer

I am not sure if we should have this test as shouldPassIfConditionClassIsPassedWithRequiredConditionProperties() sounds similar to it.

File: api/src/main/java/org/openmrs/validator/ConditionValidator.java [hunk 2]
```diff
@@ -50,5 +52,8 @@ public void validate(Object object, Errors errors) {
 		if(condition.getClinicalStatus() == null){
 			errors.rejectValue("clinicalStatus", "Condition.clinicalStatusShouldNotBeNull", "The clinical status is required");
 		}
+		if(condition.getClinicalStatus().equals(ConditionClinicalStatus.INACTIVE) && condition.getOnsetDate().compareTo(condition.getEndDate()) > 0){
+			errors.rejectValue("onsetDate","Condition.onsetDateShouldBeBeforeEndDate","Valid Onset Date is required");
+		}
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/validator/ConditionValidator.java]: This would work:
```java
if (condition.getClinicalStatus() == null){
  errors.rejectValue("clinicalStatus", "Condition.clinicalStatusShouldNotBeNull", "The clinical status is required");
}
else if (condition.getClinicalStatus().equals(ConditionClinicalStatus.INACTIVE) && condition.getOnsetDate().compareTo(condition.getEndDate()) > 0) {
  errors.rejectValue("onsetDate","Condition.onsetDateShouldBeBeforeEndDate","Valid Onset Date is required");
}
```

Comment (haripriya999) [reply api/src/main/java/org/openmrs/validator/ConditionValidator.java]: Thank you @mks-d . This fixed the issue :)

File: api/src/main/java/org/openmrs/validator/ConditionValidator.java [hunk 3]
```diff
@@ -46,9 +48,11 @@ public void validate(Object object, Errors errors) {
 		Condition condition = (Condition) object;
 		if(condition.getCondition() == null){
 			errors.rejectValue("condition", "Condition.conditionShouldNotBeNull", "The condition is required");
+		} else if(condition.getClinicalStatus().equals(ConditionClinicalStatus.INACTIVE) && condition.getOnsetDate().compareTo(condition.getEndDate()) > 0){
+			errors.rejectValue("onsetDate","Condition.onsetDateShouldBeBeforeEndDate","Valid Onset Date is required");
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/validator/ConditionValidator.java]: Can you add a test for this?

Comment (haripriya999) [reply api/src/main/java/org/openmrs/validator/ConditionValidator.java]: I added shouldPassForValidOnsetDate() to test this in ConditionValidatorTest.java

Comment (dkayiwa) [reply api/src/main/java/org/openmrs/validator/ConditionValidator.java]: How is it testing the above line?

Comment (haripriya999) [reply api/src/main/java/org/openmrs/validator/ConditionValidator.java]: I added another test function shouldFailForInvalidOnsetDate() and updated the PR :)

**Comment by haripriya999** [comment]
@dkayiwa how to fix the coveralls issue? what could have been the reason for this?

File: api/src/test/java/org/openmrs/validator/ConditionValidatorTest.java [hunk 4]
```diff
@@ -80,4 +81,26 @@ public void shouldPassIfConditionClassIsPassedWithRequiredConditionProperties(){
 		Assert.assertFalse(errors.hasFieldErrors("condition"));
 		Assert.assertFalse(errors.hasFieldErrors("clinicalStatus"));
 	}
+
+	@Test
+	public void shouldPassForValidOnsetDate(){
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/validator/ConditionValidatorTest.java]: When is an onset date valid? This should be reflected in the method name.

File: api/src/test/java/org/openmrs/validator/ConditionValidatorTest.java [hunk 5]
```diff
@@ -80,4 +81,26 @@ public void shouldPassIfConditionClassIsPassedWithRequiredConditionProperties(){
...
+		condition.setCondition(new CodedOrFreeText(new Concept(), new ConceptName("name", new Locale("en")), "nonCoded"));
+		condition.setOnsetDate(new Date(1562591017000L));
+		condition.setEndDate(new Date(1562936617000L));
+		condition.setClinicalStatus(ConditionClinicalStatus.INACTIVE);
+		validator.validate(condition,errors);
+		Assert.assertFalse(errors.hasFieldErrors("onsetDate"));
+	}
+
+	@Test
+	public void shouldFailForInvalidOnsetDate(){
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/validator/ConditionValidatorTest.java]: When is an onset date invalid? This should be reflected in the method name.

**Comment by haripriya999** [comment]
@dkayiwa  I updated the PR

**Comment by haripriya999** [comment]
@dkayiwa I don't understand why coveralls is failing on this. Could you please suggest something. Its surprising to see this as I have added tests as well.

**Comment by haripriya999** [comment]
@dkayiwa  Coveralls problem has been fixed. It occurred because of having more test cases as juliet suggested : https://talk.openmrs.org/t/coverage-coveralls-issue/22187/11 . I merged 2 tests into one test function and the issue got fixed.

**Comment by haripriya999** [comment]
@dkayiwa Coveralls issue has been fixed and is ready to be merged. Thank you :)

**Comment by haripriya999** [comment]
@mks-d  Can I  get your approval on this? Thank you :)

File: api/src/main/java/org/openmrs/validator/ConditionValidator.java (hunk 3 above)
Comment (mks-d) [reply api/src/main/java/org/openmrs/validator/ConditionValidator.java]: @haripriya999 where's this `shouldFailForInvalidOnsetDate()`? I only see `shouldPassWhenOnsetDateIsBeforeEndDate()` that seem to be testing both the 'pass' and 'fail' cases.

File: api/src/main/java/org/openmrs/validator/ConditionValidator.java [hunk 6]
```diff
@@ -46,7 +48,9 @@ public void validate(Object object, Errors errors) {
 		Condition condition = (Condition) object;
 		if(condition.getCondition() == null){
 			errors.rejectValue("condition", "Condition.conditionShouldNotBeNull", "The condition is required");
-		}
+		} else if(condition.getClinicalStatus().equals(ConditionClinicalStatus.INACTIVE) && condition.getOnsetDate().compareTo(condition.getEndDate()) > 0){
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/validator/ConditionValidator.java]: That'a detail but I personally find this a lot more readable using `Date.after(Date)`:
```java
else if (condition.getClinicalStatus().equals(ConditionClinicalStatus.INACTIVE) &&
  condition.getOnsetDate().after(condition.getEndDate())) {
```
You don't _have_ to change that, also this would push you in re-testing edge cases. But for you info.

File: api/src/main/java/org/openmrs/validator/ConditionValidator.java (hunk 3 above)
Comment (haripriya999) [reply api/src/main/java/org/openmrs/validator/ConditionValidator.java]: @mks-d  I wrote the code for both the test's in the same function (shouldPassWhenOnsetDateIsBeforeEndDate) in different blocks, as writing two separate test cases is causing a Coveralls error. Merging the two tests into one function fixed the Coveralls error.

File: api/src/main/java/org/openmrs/validator/ConditionValidator.java (hunk 6 above)
Comment (haripriya999) [reply api/src/main/java/org/openmrs/validator/ConditionValidator.java]: Thank you @mks-d . I will keep this in mind :)

File: api/src/main/java/org/openmrs/validator/ConditionValidator.java (hunk 3 above)
Comment (mks-d) [reply api/src/main/java/org/openmrs/validator/ConditionValidator.java]: Then call it `shouldValidateThatOnsetDateIsBeforeEndDate()`.

Can you still point us to the Coveralls error trace?

Also don't force push, just add commits to your PR.

Comment (haripriya999) [reply api/src/main/java/org/openmrs/validator/ConditionValidator.java]: Thank you @mks-d .Pushing it as separate commits isn't causing the coveralls issue even after putting the code in separate function i.e shouldFailForInvalidOnsetDate() .  Please review the changes. :)

**Comment by gitcliff** [comment]
@haripriya999 can you look into the merge conflicts ..thanks
